coreneuron/permute/cellorder.cu

```
@@ -0,0 +1,99 @@
+        nt->_actual_rhs[i] /= nt->_actual_d[i];  // the root
+    }
+
+    int i = firstnode + icore;
+
+    int ip;
+    for (int icycle = 0; icycle < ncycle; ++icycle) {
+        int istride = stride[icycle];
+        if (icore < istride) {
+            ip = nt->_v_parent_index[i];
+            nt->_actual_rhs[i] -= nt->_actual_b[i] * nt->_actual_rhs[ip];
+            nt->_actual_rhs[i] /= nt->_actual_d[i];
+        }
+        i += istride;
+    }
+}
+
+__global__ void solve_interleaved2_kernel(NrnThread* nt, InterleaveInfo* ii, int ncore) {
+    int icore = blockDim.x * blockIdx.x + threadIdx.x;
+
+    int* ncycles = ii->cellsize;         // nwarp of these
+    int* stridedispl = ii->stridedispl;  // nwarp+1 of these
+    int* strides = ii->stride;           // sum ncycles of these (bad since ncompart/warpsize)
+    int* rootbegin = ii->firstnode;      // nwarp+1 of these
+    int* nodebegin = ii->lastnode;       // nwarp+1 of these
+
+    int iwarp = icore / warpsize;     // figure out the >> value
+    int ic = icore & (warpsize - 1);  // figure out the & mask
+    int ncycle = ncycles[iwarp];
+    int* stride = strides + stridedispl[iwarp];
+    int root = rootbegin[iwarp];
+    int lastroot = rootbegin[iwarp + 1];
+    int firstnode = nodebegin[iwarp];
+    int lastnode = nodebegin[iwarp + 1];
+
+    triang_interleaved2_device(nt, ic, ncycle, stride, lastnode);
+    bksub_interleaved2_device(nt, root + ic, lastroot, ic, ncycle, stride, firstnode);
+}
+
+void solve_interleaved2_launcher(NrnThread* nt, InterleaveInfo* info, int ncore, void* stream) {
+    auto cuda_stream = static_cast<cudaStream_t>(stream);
+
+    int threadsPerBlock = warpsize;
+    // TODO: Should blocksPerGrid be a fixed number and have a while block inside the kernel?
+    int blocksPerGrid = (ncore + threadsPerBlock - 1) / threadsPerBlock;
+
+    solve_interleaved2_kernel<<<blocksPerGrid, threadsPerBlock, 0, cuda_stream>>>(nt, info, ncore);
+
+    cudaStreamSynchronize(cuda_stream);
```

@kotsaloscv : Could we do error checking here? to make sure the kernels are executed and avoid silent errors...

---

@pramodk : I just added a check. Shall we move the check in the utils folder for future CUDA-related development?

---

Sure! sounds good! (realised `coreneuron/utils/randoms/nrnran123.cu` also has some cuda calls without error checking)

---

regarding `coreneuron/utils/randoms/nrnran123.cu` shall I open a ticket or include a fix in this PR?

---

lets do that separately. Go ahead with a ticket and then one can look at other parts of the code as well.